go/vt/vtgate/engine/insert.go

```diff
@@ -174,9 +174,20 @@ func (ins *Insert) executeInsertQueries(
 		return nil, vterrors.Aggregate(errs)
 	}
 
+	// If this insert used auto increment values from a sequence, we need to set the `last_insert_id` value.
 	if insertID != 0 {
-		result.InsertID = insertID
+		if result.RowsAffected > 0 {
+			// If at least one row was affected, we set the `last_insert_id` value to the lowest reserved sequence id.
+			//
+			// This does not match the behaviour of MySQL in case where no new rows where inserted (but one or more rows were updated
+			// via `ON DUPLICATE KEY UPDATE`), where the `last_insert_id` value is set to the `auto_increment` column value.
+			result.InsertID = insertID
+		} else {
+			// If no rows were inserted or updated, clear the `last_insert_id` value.
+			result.InsertID = 0
+		}
 	}
+
```

Comment:
this is not true. I tried it on mysql

```mysql
mysql> insert into at(col) values (20);
Query OK, 1 row affected (0.00 sec)

mysql> select * from at where col = 20;
+----+------+
| id | col  |
+----+------+
| 10 |   20 |
+----+------+
1 row in set (0.00 sec)

mysql> insert into at(col) values (20);
ERROR 1062 (23000): Duplicate entry '20' for key 'at.col'

mysql> insert into at(col) values (20) on duplicate key update col = 40;
Query OK, 2 rows affected (0.00 sec)

mysql> select * from at where col = 40;
+----+------+
| id | col  |
+----+------+
| 10 |   40 |
+----+------+
1 row in set (0.00 sec)

mysql> select @@last_insert_id;
+------------------+
| @@last_insert_id |
+------------------+
|               10 |
+------------------+
1 row in set (0.00 sec)
```

If you follow through this sample test. The last_insert_id remained the old value of `10` which was the last insert id when `20` for `col` was inserted into the table.